armory/art_experimental/attacks/dapricot_patch.py

```diff
@@ -1,23 +1,673 @@
+            isinstance(s, float) for s in self.brightness_range
+        ):
+            raise ValueError(
+                "The brightness range must be either a tuple or list of floats."
+            )
+        if len(self.brightness_range) != 2:
+            raise ValueError("The length of brightness range must be 2.")
+
+        # if self.brightness_range[0] < 0.0 or self.brightness_range[1] > 1.0:
+        #     raise ValueError("The brightness range must be between 0.0 and 1.0.")
+
+        if self.brightness_range[0] > self.brightness_range[1]:
+            raise ValueError(
+                "The first element of the brightness range must be less or equal to the second one."
+            )
+
+        if not isinstance(self.rotation_weights, (tuple, list)) or not all(
+            isinstance(s, (float, int)) for s in self.rotation_weights
+        ):
+            raise ValueError(
+                "The rotation sampling weights must be provided as tuple or list of float or int values."
+            )
+        if len(self.rotation_weights) != 4:
+            raise ValueError("The number of rotation sampling weights must be 4.")
+
+        if not all(s >= 0.0 for s in self.rotation_weights):
+            raise ValueError("The rotation sampling weights must be non-negative.")
+
+        if all(s == 0.0 for s in self.rotation_weights):
+            raise ValueError(
+                "At least one of the rotation sampling weights must be strictly greater than zero."
+            )
+
+        if not isinstance(self.sample_size, int):
+            raise ValueError("The EOT sample size must be of type int.")
+        if self.sample_size <= 0:
+            raise ValueError("The EOT sample size must be greater than 0.")
+
+        if not isinstance(self.targeted, bool):
+            raise ValueError("The argument `targeted` has to be of type bool.")
+
+    def _augment_images_with_patch(self, x, y, patch, channels_first):
+        """
+        Augment images with patch using perspective transform
+        instead of inserting patch based on patch_location
+
+        :param x: Sample images.
+        :param y: Target labels.
+        :param patch: The patch to be applied.
+        :param channels_first: Set channels first or last.
+        """
+
+        if x.shape[0] != len(self.gs_coords):
+            raise ValueError(
+                "Number of images should be equal to the number of arrays of green screen coordinates"
+            )
+        if y is not None and (x.shape[0] != len(y)):
+            raise ValueError(
+                "Number of images should be equal to the number of targets"
+            )
+
+        transformations = dict()
+        x_copy = x.copy()
+        patch_copy = patch.copy()
+        x_patch = x.copy()
+
+        if channels_first:
+            x_copy = np.transpose(x_copy, (0, 2, 3, 1))
+            x_patch = np.transpose(x_patch, (0, 2, 3, 1))
```

Comment:
`x_patch` as defined by `x_patch = np.transpose(x_patch, (0, 2, 3, 1))` will not be used as this variable is then reset to an empty list. Should this line be removed?

Reply:
Wouldn't this be used if the model is Pytorch?

Reply:
We should then remove any code under `if channels_first:`

Reply:
My point here was just that `x_patch` is then, regardless, immediately set to an empty list in [this](https://github.com/yusong-tan/armory/blob/1039-dapricot-scale_ytan/armory/art_experimental/attacks/dapricot_patch.py#L518) line, making [this](https://github.com/yusong-tan/armory/blob/1039-dapricot-scale_ytan/armory/art_experimental/attacks/dapricot_patch.py#L514) line irrelevant. 

But I think you're right in that we should remove code under `if channels_first`, considering the Armory always loads data in channels_last format.

Reply:
I agree. I did not see which line you were referring to when you said "then reset to an empty list," but now I know.